lib/ui/ui_utils.dart

```diff
@@ -84,3 +55,82 @@ CoreElement createHotReloadButton() {
   });
   return button;
 }
+
+class ServiceExtensionButton {
+  ServiceExtensionButton(this.extensionDescription) {
+    button = new PButton.icon(
+        extensionDescription.description, extensionDescription.icon,
+        title: extensionDescription.tooltip)
+      ..small();
+
+    final extensionName = extensionDescription.extension;
+
+    // Disable button for unavailable service extensions.
+    button.disabled = !serviceManager.serviceExtensionManager
+        .isServiceExtensionAvailable(extensionName);
+    serviceManager.serviceExtensionManager.hasServiceExtension(
+        extensionName, (available) => button.disabled = !available);
+
+    button.click(() => click());
+
+    manageState();
+  }
+
+  final ServiceExtensionDescription extensionDescription;
+  PButton button;
+
+  void click() {
+    final bool wasSelected = button.element.classes.contains('selected');
+    serviceManager.serviceExtensionManager.setServiceExtensionState(
+      extensionDescription.extension,
+      !wasSelected,
+      wasSelected
+          ? extensionDescription.disabledValue
+          : extensionDescription.enabledValue,
+    );
+  }
+
+  void manageState() {
```

Comment:
nit: manageState seems like a little bit of an odd name.
Also, make this private to clarify users shouldn't calll it.
Perhaps: `_updateState`

Reply:
done.